File: binarytree/145_postorder_traversal.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # def postorderTraversal(self, root: TreeNode) -> List[int]:
    #     if not root:
    #         return []
    #     return [
    #         *self.postorderTraversal(root.left),
    #         *self.postorderTraversal(root.right),
    #         root.val
    #     ]

    # def postorderTraversal(self, root: TreeNode) -> List[int]:
    #     if not root:
    #         return []
    #     nums: List[int] = []
    #     stack: List[TreeNode] = [root]
    #     while stack:
    #         node = stack.pop()
    #         nums.append(node.val)
    #         if node.left:
    #             stack.append(node.left)
    #         if node.right:
    #             stack.append(node.right)
    #     nums.reverse()
    #     return nums

    def postorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        stack: List[TreeNode] = []
        
        prev: Optional[TreeNode] = None
        curr: Optional[TreeNode] = root
        nums: List[int] = []
        while curr or stack:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            if not curr.right or curr.right == prev:
                nums.append(curr.val)
                prev, curr = curr, None
            else:
                stack.append(curr)
                curr = curr.right
        return nums



if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    res = Solution().postorderTraversal(root)
    print(res)

File: binarytree/145_postorder_traversal.cc
#include <stack>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ret;
        if (root == nullptr) {
            return ret;
        }

        TreeNode* prev = nullptr;
        TreeNode* curr = root;
        stack<TreeNode*> _stack;
        while (curr != nullptr || !_stack.empty())
        {
            while (curr != nullptr)
            {
                _stack.push(curr);
                curr = curr->left;
            }
            curr = _stack.top();
            _stack.pop();
            if (curr->right == nullptr || curr->right == prev) {
                ret.push_back(curr->val);
                prev = curr;
                curr = nullptr;
                // prev, curr = curr, nullptr;
            } else {
                _stack.push(curr);
                curr = curr->right;
            }
        }
        return ret;
    }
};

int main() {
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    vector<int> ret = Solution().postorderTraversal(root);
    for_each(ret.begin(), ret.end(), [](int &i){ printf("%d\n", i); });
    return 0;
}